### PR TITLE
[Feature/Identity] Rest APIs for CRUD actions Internal users

### PR DESCRIPTION
**Pull request by DarshitChanpura**
P.S: this PR is based off of https://github.com/opensearch-project/OpenSearch/pull/5613. I will keep rebasing with it till it is merged.

### Description
Sets up base for REST API architecture for identity project

### Issues Resolved
TBD

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java [hunk 1]
```diff
@@ -61,4 +61,19 @@ public void setAttributes(Map<String, String> attributes) {
     public String toString() {
         return "User [username=" + username + ", bcryptHash=" + bcryptHash + ", attributes=" + attributes + "]";
     }
+
+    @Override
+    public String getHash() {
```
Comment (cwperks) [inline sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:66]: Why not rename the other methods `getBcryptHash` and `setBcryptHash` to `getHash` and `setHash` respectively?

File: sandbox/modules/identity/build.gradle [hunk 2]
```diff
@@ -22,6 +22,7 @@ dependencies {
   implementation project(':sandbox:libs:opensearch-authn')
 
   implementation 'org.greenrobot:eventbus:3.2.0'
+  implementation 'com.google.guava:guava:31.1-jre'
```
Comment (cwperks) [inline sandbox/modules/identity/build.gradle:25]: Guava is a forbidden API in opensearch: https://github.com/opensearch-project/OpenSearch/blob/main/gradle/forbidden-dependencies.gradle#L15-L17

Alternative data structures should be used. I'm not quite sure of the suggested replacements for ImmutableSet and ImmutableList.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Hideable.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.configuration;
+
+public interface Hideable {
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Hideable.java:11]: Let's add these only when needed. There is no concept of hidden or reserved at the moment.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/SecurityDynamicConfiguration.java [hunk 4]
```diff
@@ -125,6 +132,39 @@ public T getCEntry(String key) {
         return centries.get(key);
     }
 
+    @JsonIgnore
+    public void removeHidden() {
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/SecurityDynamicConfiguration.java:136]: Similar to above, do we need these `removeHidden` and `removeStatic`?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+
+        if (!isWriteable(channel, existingConfiguration, name)) {
+            return;
+        }
+
+        boolean existed = existingConfiguration.exists(name);
+        existingConfiguration.remove(name);
+
+        if (existed) {
+            saveAnUpdateConfigs(
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java:155]: nit: `saveAndUpdateConfigs`

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+        final String name = request.param("name");
+
+        if (name == null || name.length() == 0) {
+            badRequestResponse(channel, "No " + getResourceName() + " specified.");
+            return;
+        }
+
+        final SecurityDynamicConfiguration<?> existingConfiguration = load(getConfigName());
+
+        if (existingConfiguration.getSeqNo() < 0) {
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java:186]: What is this block doing?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+                "Security index need to be updated to support '" + getConfigName().toLCString() + "'. Use SecurityAdmin to populate."
+            );
+            return;
+        }
+
+        if (!isWriteable(channel, existingConfiguration, name)) {
+            return;
+        }
+
+        if (isReadonlyFieldUpdated(existingConfiguration, content)) {
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java:198]: I see that this function currently returns only `false`. How will this eventually be used?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestInternalUsersApiAction.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        final SecurityDynamicConfiguration<?> internalUsersConfiguration = load(getConfigName());
+
+        if (!isWriteable(channel, internalUsersConfiguration, username)) {
+            return;
+        }
+
+        final ObjectNode contentAsNode = (ObjectNode) content;
+        final JsonNode userContentAsNode = contentAsNode;
+
+        // TODO: Check role mapping and its validity here
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestInternalUsersApiAction.java:119]: This TODO can be removed

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestInternalUsersApiAction.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        final String hashedPassword = userContentAsNode.get("hash").asText();
+
+        final boolean userExisted = internalUsersConfiguration.exists(username);
+
+        // when updating an existing user password hash can be blank, which means no
+        // changes
+
+        // sanity checks, hash is mandatory for newly created users
+        if (!userExisted && hashedPassword == null) {
+            badRequestResponse(channel, "Please specify either 'hash' or 'password' when creating a new internal user.");
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestInternalUsersApiAction.java:131]: I think this response should only say `'hash'`, based on line 122 it does not look for a field called `password`.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestApiPrivilegesEvaluator.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.rest.api;
+
+import org.opensearch.rest.RestRequest;
+
+public class RestApiPrivilegesEvaluator {
+    public String checkAccessPermissions(RestRequest request, Endpoint endpoint) {
+        // TODO: implement this in future scope of Identity
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestApiPrivilegesEvaluator.java:15]: 👍

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+
+        // check if .opendistro_security index has been initialized
+        if (!ensureIndexExists()) {
+            return channel -> internalErrorResponse(channel, ErrorType.SECURITY_NOT_INITIALIZED.getMessage());
+        }
+
+        // check if request is authorized
+        String authError = restApiPrivilegesEvaluator.checkAccessPermissions(request, getEndpoint());
+
+        final User user = (User) threadPool.getThreadContext().getTransient(ConfigConstants.IDENTITY_USER);
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java:388]: This is where we will want to use the `Identity.getAuthManager().getSubject()` and abstract this logic out.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+        final String userName = user == null ? null : user.getUsername().getName();
+        if (authError != null) {
+            log.error("No permission to access REST API: " + authError);
+            // for rest request
+            request.params().clear();
+            return channel -> forbidden(channel, "No permission to access REST API: " + authError);
+        }
+
+        final Object originalUser = threadPool.getThreadContext().getTransient(ConfigConstants.IDENTITY_USER);
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/IdentityAbstractRestApiAction.java:397]: @peternied Any thoughts on how interactions with ThreadContext should work in the identity module?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/RoutesHelper.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+public class RoutesHelper {
+
+    /**
+     * Add prefixes(_opendistro... and _plugins...) to rest API routes
+     * @param routes routes
+     * @return new list of API routes prefixed with _opendistro... and _plugins...
+     *Total number of routes is expanded as twice as the number of routes passed in
+     */
+    public static List<RestHandler.Route> addRoutesPrefix(List<RestHandler.Route> routes) {
+        return addRoutesPrefix(routes, "/_opendistro/_security/api", "/_plugins/_security/api");
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/RoutesHelper.java:25]: These should be changed.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/RoutesHelper.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+            .collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Add prefixes(_plugins...) to rest API routes
+     * @param deprecatedRoutes Routes being deprecated
+     * @return new list of API routes prefixed with _opendistro... and _plugins...
+     *Total number of routes is expanded as twice as the number of routes passed in
+     */
+    public static List<RestHandler.DeprecatedRoute> addDeprecatedRoutesPrefix(List<RestHandler.DeprecatedRoute> deprecatedRoutes) {
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/RoutesHelper.java:47]: Remove this?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/validation/CredentialsValidator.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+     * Function to validate password in the content body.
+     * @return true if validation is successful else false
+     */
+    @Override
+    public boolean validate() {
+        if (!super.validate()) {
+            return false;
+        }
+
+        final String regex = this.opensearchSettings.get(ConfigConstants.IDENTITY_RESTAPI_PASSWORD_VALIDATION_REGEX, null);
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/validation/CredentialsValidator.java:46]: wdyt about providing a default? Does it make sense to have this be configurable?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/validation/InternalUsersValidator.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+    public InternalUsersValidator(
+        final RestRequest request,
+        boolean isSuperAdmin,
+        BytesReference ref,
+        final Settings opensearchSettings,
+        Object... param
+    ) {
+        super(request, ref, opensearchSettings, param);
+        allowedKeys.put("backend_roles", DataType.ARRAY);
```
Comment (cwperks) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/validation/InternalUsersValidator.java:28]: Can `backend_roles` and `opendistro_security_roles` be removed here?

**Review by cwperks — COMMENTED** [review]
Thank you @DarshitChanpura, this provides a good base to build upon. I think we should reduce code down to core elements where we can to minimize the change and create an issue to specify testing scenarios to build test cases on for updating a user via the API.

**Review by peternied — CHANGES_REQUESTED** [review]
API actions should not be implemented like this, they should follow the latest convention from OpenSearch, checkout how Delete for Point In Time was [implemented](https://github.com/search?q=repo%3Aopensearch-project%2FOpenSearch%20DeletePitAction&type=code) in these places.

There should be a class 'UserService' that knows how to interact with the internal systems to perform these actions and manage state, there is an analog to the [PITService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/PitService.java).  That class would also isolate all configuration updates/propagation, and on this note I would focus this PR on declaring the minimal REST APIs / Service requirements and leave todos around more advanced features like concurrent optmization.

**Comment by DarshitChanpura** [comment]
> API actions should not be implemented like this, they should follow the latest convention from OpenSearch, checkout how Delete for Point In Time was [implemented](https://github.com/search?q=repo%3Aopensearch-project%2FOpenSearch%20DeletePitAction&type=code) in these places.
> 
> There should be a class 'UserService' that knows how to interact with the internal systems to perform these actions and manage state, there is an analog to the [PITService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/PitService.java). That class would also isolate all configuration updates/propagation, and on this note I would focus this PR on declaring the minimal REST APIs / Service requirements and leave todos around more advanced features like concurrent optmization.

As an alternative to this PR, I have created another PR #5990 which leverages the way that core implements REST APIs.

**Comment by DarshitChanpura** [comment]
Closing this as it is not actively worked on

**Pull request closed without merging**